Add channel specific application instances

dylanmccall opened the pull request:
Closes #28

To create a channel specific application instance, use the `--channel-id` argument:

```
kolibri-gnome --channel-id=c9d7f950ab6b5a1199e3d6c10d7f0103
```

This will work better if you have a desktop file `~/.local/share/applications/org.learningequality.Kolibri.Devel.channel_c9d7f950ab6b5a1199e3d6c10d7f0103.desktop` with contents like the following:

```ini
[Desktop Entry]
Version=1.0
Type=Application
Name=Khan Academy (English - US curriculum)
Comment=
Exec=gio open "kolibri:c9d7f950ab6b5a1199e3d6c10d7f0103"
X-Endless-LaunchMaximized=True
X-Kolibri-Channel-Id=c9d7f950ab6b5a1199e3d6c10d7f0103
X-Kolibri-Channel-Version=2~2
Categories=Education;
Icon=/var/home/dylan/.var/app/org.learningequality.Kolibri.Devel/data/kolibri/content/xdg/share/icons/org.learningequality.Kolibri.Channel.c9d7f950ab6b5a1199e3d6c10d7f0103.png
```

That desktop file will not open the channel specific application on its own (that is a problem for another branch), but it will cause GNOME Shell to consistently identify it with the correct name and icon.

Remaining tasks:

 - [x] Set a different app ID for standalone mode, matching the desktop files generated by https://github.com/endlessm/kolibri-app-desktop-xdg-plugin. Ensure this is running in a separate process.
 - [x] When opening a Kolibri URL that is unsupported by the channel application, open it in the generic Kolibri application instead of a browser window.
- [x] Reorganize some hacky code.

dylanmccall committed:
Add standalone windows which are restricted to specific URLs

Issue #28

dylanmccall committed:
Use a different Application for standalone mode

This includes a different application ID so different channels can
appear as different applications in the user's desktop environment.

Issue #28

dylanmccall committed:
Move standalone window code into ChannelApplication

Issue #28

dylanmccall committed:
Add navigation buttons to the header bar

In addition, enable and disable them based on the context.

Issue #28

dylanmccall committed:
Use Kolibri's API to limit content in standalone windows

Issue #28

danigm commented on src/kolibri_gnome/desktop_launcher/application.py:
```diff
@@ -491,3 +695,100 @@ def open_kolibri_home(self):
+
+        if re.match(r"^\/(?P<lang>\w+\/)?learn\/?", url_tuple.path):
+            return self.__is_learn_fragment_in_channel(url_tuple.fragment)
+        elif re.match(r"^\/(?P<lang>\w+\/)?user\/?", url_tuple.path):
+            return True
+        elif re.match(r"^\/(?P<lang>\w+\/)?logout\/?", url_tuple.path):
+            return True
+        elif re.match(r"^\/(?P<lang>\w+\/)?redirectuser\/?", url_tuple.path):
+            return True
+        elif re.match(r"^\/zipcontent\/?", url_tuple.path):
+            return True
+        elif re.match(r"^\/app\/?", url_tuple.path):
+            return True
+        elif re.match(r"^\/static\/?", url_tuple.path):
+            return True
+        elif re.match(r"^\/downloadcontent\/?", url_tuple.path):
+            return True
+        elif re.match(r"^\/content\/storage\/?", url_tuple.path):
```
This can be simplified grouping regular expressions:
```suggestion
        elif re.match(r"^\/(?P<lang>\w+\/)?(user|logout|redirectuser)\/?", url_tuple.path):
            return True
        elif re.match(r"^\/(zipcontent|app|static|downloadcontent|content\/storage)\/?", url_tuple.path):
            return True
```

But maybe this way is clearer.

dylanmccall replied:
Ah, I like it. I'd started off copying these from Kolibri's urls.py, but seeing as this code is nowhere near that code, the similarity between the two doesn't really help anything. That's much better! Thank you :)

danigm commented on src/kolibri_gnome/desktop_launcher/application.py:
```diff
@@ -491,3 +695,100 @@ def open_kolibri_home(self):
+        return contentnode_channel == self.channel_id
+
+    def __contentnode_id_for_learn_fragment(self, fragment):
+        patterns = (
+            r"^topics\/c\/(?P<node_id>\w+)",
+            r"^topics\/t\/(?P<node_id>\w+)",
+            r"^topics\/(?P<node_id>\w+)",
+        )
+
+        for pattern in patterns:
+            match = re.match(pattern, fragment)
+            if match:
+                return match.group("node_id")
```
This is something that could also be simplified using just one regular expression:
```suggestion
        pattern = r"^topics\/([ct]\/)?(?P<node_id>\w+)"
        match = re.match(pattern, fragment)
        if match:
            return match.group("node_id")
```

danigm reviewed (approved):
I've been testing and it's working for me. I've just left two nitpick comments about regular expressions that can be combined in one.

dylanmccall committed:
Apply suggestions from code review

https://phabricator.endlessm.com/T31639

Co-authored-by: danigm <dani@danigm.net>

danigm committed:
Fix regular expression to support compound lang names

The lang can be something like 'en-us' so it's needed to handled that in
the regular expression.

https://phabricator.endlessm.com/T31639

danigm committed:
Add new kolibri-launcher app

This new app is a simple Gio.Application that handles the
kolibri-channel uri scheme and launch the kolibri-gnome with the
channel-id parameter.

This new launcher could be useful to launch ChannelApplication directly
using `gio open kolibri-channel:3e464ee12f6a50a781cddf59147b48b1`. This
is not possible with the Pew app because by default it will launch the
GenericApplication so this utility launcher will just handle the scheme
and call the real app with the corresponding parameters.

danigm commented on src/kolibri_gnome/desktop_launcher/application.py:
```diff
@@ -491,3 +695,83 @@ def open_kolibri_home(self):
+            return True
+        elif re.match(r"^\/(zipcontent|app|static|downloadcontent|content\/storage)\/?", url_tuple.path):
+            return True
+            return True
```
This line is duplicated and can be removed.

```suggestion
```

dylanmccall committed:
Update src/kolibri_gnome/desktop_launcher/application.py

Co-authored-by: danigm <dani@danigm.net>

dylanmccall committed:
Add kolibri-explore-plugin URLs to regex

danigm committed:
Split desktop_launcher module

This patch splits the desktop_launcher python module in two different
modules, the "desktop" and the "launcher" modules.

The "desktop" module contains the whole gtk application and the
"launcher" one just contain the "kolibri-channel" gio scheme launcher.
In the future we can move all the gio schemes management to the launcher
application and pass all the information to the "desktop" application
using command line arguments.

dylanmccall commented:
This seems to have been working well, so I'll merge this into the master branch now.
